static/app/components/carousel.tsx

```diff
@@ -0,0 +1,148 @@
+import {useEffect, useRef, useState} from 'react';
+import styled from '@emotion/styled';
+
+import {Button} from 'sentry/components/button';
+import {IconArrow} from 'sentry/icons';
+
+type Props = {
+  children?: React.ReactNode;
+};
+
+function Carousel({children}: Props) {
+  const ref = useRef<HTMLDivElement | null>(null);
+  const [isAtStart, setIsAtStart] = useState<boolean>(true);
+  const [isAtEnd, setIsAtEnd] = useState<boolean>(false);
+  const [showArrows, setShowArrows] = useState<boolean>(true);
+
+  const setStartAndEnd = (scrollLeft: number, totalScrollDist: number) => {
+    // scrollLeft: the amount scrolled horizontally.
+    // start at 0 and end at totalScrollDist
+
+    // totalScrollDist = scrollWidth - clientWidth
+    // scrollWidth: total element width including overflow
+    // clientWidth: width of the element excluding overflow
+    if (scrollLeft <= 0) {
+      setIsAtStart(true);
+      setIsAtEnd(false);
+    } else if (scrollLeft >= totalScrollDist) {
+      setIsAtStart(false);
+      setIsAtEnd(true);
+    } else {
+      setIsAtStart(false);
+      setIsAtEnd(false);
+    }
+  };
```

Comment:
Rather than calling `setStartAndEnd()` on scroll, I'd recommend using [`IntersectionObservers`](https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API) with two invisible, zero-width anchors, one at the start and another at the end of the carousel. If the start anchor is in view (`entry.isIntersecting` is true), then we're at the start. Same for the end anchor. The main advantage of `IntersectionObservers` is efficiency — the callback function is only called when an anchor moves into/out of view.

Reply:
I don't even think you would need the anchors, I think you could use the `threshold`

Reply:
That said we would probably want to then export a component named `CarouselItem` that uses context to automatically propagate the ref upwards to this component.

The type of children would then be `React.ReactElement<CarouselItemProps>` or something like that